File: P1S1/src/P1/Usuario.java

package P1;

/**
 *
 * Estudiantes: Óscar Antunez Matinaitis y Amanda Moyano Romero
 * Grupo DS2 12
 */

import java.util.Observable;
import java.util.Observer;
import java.util.ArrayList;

public class Usuario extends Observable {

	private String nombreUsuario;
    private ArrayList<Usuario> seguidores;
    private ArrayList<Usuario> seguidos;
    
    public Usuario(String nombreUsuario){
        this.nombreUsuario = nombreUsuario;
        seguidores = new ArrayList<>();
        seguidos = new ArrayList<>();
    }

	/**
	 * 
	 * @param aSeguir
	 */
	public void seguir(Usuario aSeguir) {
            if(!seguidos.contains(aSeguir)){
                seguidos.add(aSeguir);
                aSeguir.addSeguidor(this);
            }
	}

    private synchronized void addSeguidor(Usuario seguidor){
        seguidores.add(seguidor);

        setChanged();
        notifyObservers();
    }

    public synchronized void addMonitor(Observer monitor){
        addObserver(monitor);
    }

    public ArrayList<Usuario> getSeguidores(){
        return new ArrayList<Usuario>(seguidores);
    }

    public String getNombre(){
        return nombreUsuario;
    }

}

File: P2/lib/modelo/Usuario.dart
class Usuario {
  String _nombreUsuario;
  String _password;
  List<Usuario> _seguidos;
  List<Usuario> _seguidores;
  List<String> _busquedasRecientes;

  String getNombre(){
    return _nombreUsuario;
  }

  String getPassword(){
    return _password;
  }

  List<Usuario> getSeguidores(){
    return _seguidores;
  }

  List<Usuario> getSeguidos(){
    return _seguidos;
  }

  bool isSeguido(Usuario usu){
    if(this._seguidos.contains(usu))
      return true;

    else
      return false;
  }

  bool isSeguidor(Usuario usu){
    if(usu.isSeguido(this))
      return true;

    else
      return false;
  }

  void seguir(Usuario u){
    if(!isSeguido(u) && u != this)
      _seguidos.add(u);
  }

  void addSeguidor(Usuario u){
    if(!_seguidores.contains(u) && u != this)
      _seguidores.add(u);
  }

  Usuario(String nombreUsuario, String password){
    this._nombreUsuario = nombreUsuario;
    this._password = password;
    _seguidos = [];
    _seguidores = [];
    _busquedasRecientes = [];
  }

  List<String> getBusquedasRecientes(){
    return _busquedasRecientes;
  }

  void pushBusquedasRecientes(String query){
    _busquedasRecientes.add(query);
  }
}